services/wallet/controllers_v3.go

```diff
@@ -441,49 +449,144 @@ func LinkUpholdDepositAccountV3(s *Service) func(w http.ResponseWriter, r *http.
 	}
 }
 
-// GetWalletV3 - produces an http handler for the service s which handles getting of brave wallets
+const errOriginForbidden model.Error = "request origin forbidden"
+
+type linkSolanaAddrRequest struct {
+	SolanaPublicKey string `json:"solanaPublicKey" valid:"length(32|44)"`
+	Message         string `json:"message" valid:"required"`
+	SolanaSignature string `json:"solanaSignature" valid:"required"`
+}
+
+func LinkSolanaAddress(s *Service) handlers.AppHandler {
+	return func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		ctx := r.Context()
+
+		if dis, ok := ctx.Value(appctx.DisableGeminiLinkingCTXKey).(bool); ok && dis {
+			return handlers.ValidationError("Connecting Brave Rewards to Solana is temporarily unavailable. Please try again later", nil)
+		}
+
+		l := logging.Logger(ctx, "wallet")
+
+		if o := r.Header.Get("Origin"); o != s.dappConf.AllowedOrigin {
+			l.Error().Err(errOriginForbidden).Str("origin", strOr(o, "empty")).Msg("error linking solana address")
+			return handlers.WrapError(errOriginForbidden, "request origin forbidden", http.StatusForbidden)
+		}
+
+		var paymentID inputs.ID
+		if err := inputs.DecodeAndValidateString(ctx, &paymentID, chi.URLParam(r, "paymentID")); err != nil {
+			return handlers.WrapError(err, "invalid paymentID", http.StatusBadRequest)
+		}
+
+		b, err := io.ReadAll(io.LimitReader(r.Body, reqBodyLimit10MB))
```

Comment:
This is the default we are using on other places so has been added here

Reply:
@kdenhartog The only requirement is that there is some limit to the HTTP body to prevent a client or server sending gigabytes and causing memory exhaustion. Using some default like 10 is fine.